File: Backtracking/Knight's_Tour.py
'''
Aim: To check whether it is possible for a Knight to visit each
         cell of the N*N chessboard without visiting any cell
         twice starting from (X, Y) position.
Intution: To visit each and every positions which are available from
          the current position and recursively repeat this until
          all the cells are covered.
        
'''

from time import time


def move_Knight(n, pos, move, ans, step):
    x, y = pos

    # Base Case
    if step == n * n:
        return ans
    for i in move:
        r = x + i[0]
        c = y + i[1]
        if 0 <= r < n and 0 <= c < n and not ans[r][c]:
            ans[r][c] = step + 1
            temp = move_Knight(n, (r, c), move, ans, step + 1)
            if temp is not False:
                return temp
            ans[r][c] = 0
    return False


def Knight_Tour(n, pos):
    x, y = pos

    # All valid moves that a Knight can make
    move = [[2, 1], [2, -1], [-2, 1], [-2, -1],
            [1, 2], [1, -2], [-1, 2], [-1, -2]]

    # To keep a track of already visited cells and
    # Answer Matrix
    answer = [[0]*n for i in range(n)]

    # To mark (X, Y) cell as visited
    answer[x][y] = 1

    return move_Knight(n, pos, move, answer, 1)


# ------------------------DRIVER CODE ------------------------

if __name__ == "__main__":

    # Input the initial Position of the Knight
    N = int(input("Enter the size of the Chessboard: "))
    X, Y = map(int, input("Enter Initial Position of the Knight: ").split())
    start = time()
    ans_mat = Knight_Tour(N, (X-1, Y-1))
    if ans_mat is False:
        print("Knight's Tour form the given initial position is not possible")
    else:
        print("The desired Knight's Tour :")
        for i in ans_mat:
            print("\t\t", *i)
    print("Time taken: ", time()-start)

'''

Sample Working:
    
Enter the size of the Chessboard: 5
Enter the Initial Position of the Knight: 1 1
The desired Knight's Tour :
                1 6 15 10 21
                14 9 20 5 16
                19 2 7 22 11
                8 13 24 17 4
                25 18 3 12 23
Time taken:  0.10171318054199219

COMPLEXITY:
    
     Time Complexity ->  O(8^N)
     Space Complexity -> O(N^2)

'''

File: Backtracking/Knight's_Tour.java
/*Moving according to the rules of chess knight must visit each square exactly once. Print the order of each the cell in which they are visited.*/
import java.util.*;
public class MyClass {
    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        // provide start position
        int r = sc.nextInt();
        int c = sc.nextInt();
        int board[][] = new int[n][n];
        // The 8 possible moves where a knight can move from its current position
        int posx[] = {-1,-1,1,1,-2,-2,2,2};
        int posy[] = {-2,2,-2,2,-1,1,-1,1};
        path(board, r, c, 1, posx, posy);
    }
    public static boolean path(int board[][], int row, int col, int move, int posx[], int posy[]) {
        board[row][col] = move;
        // if all the positions are visited then print the result
        if (move == board.length*board.length) {
            // printing result in matrix form
            for (int i = 0; i < board.length; i++) {
                for (int j = 0; j < board.length; j++)
                    if (board[i][j] < 10)
                        System.out.print("0" + board[i][j] + " ");
                    else
                        System.out.print(board[i][j] + " ");
                System.out.println();
            }
            return true;
        }
        for (int i = 0; i < 8; i++) {
            int newx = row + posx[i];
            int newy = col + posy[i];
            // checking if the new position is a valid move and not visited yet
            if (validmove(newx, newy, board) && board[newx][newy] == 0) {
                if (path(board, newx, newy, move + 1, posx, posy)) {
                    return true;
                }
            }
        }
        // if position is not valid mark position as unvisited in the matrix
        board[row][col] = 0;
        return false;
    }
    // checks that the new position is not outside the matrix of n*n
    static boolean validmove(int newx, int newy, int board[][]) {
        if (newx >= 0 && newx < board.length && newy >= 0 && newy < board.length)
            return true;
        return false;
    }
}

/*Input
8
0 
0
Output
01 08 15 22 03 10 19 64 
16 23 02 09 18 21 04 11 
07 14 17 24 05 12 63 20 
34 25 06 13 44 27 52 61 
55 46 33 26 53 62 43 28 
32 35 54 45 40 29 60 51 
47 56 37 30 49 58 39 42 
36 31 48 57 38 41 50 59 */